File: play-akka-http-ws-standalone/src/main/scala/play/api/libs/ws/akkahttp/StandaloneAkkaHttpWSRequest.scala
/*
 * Copyright (C) 2009-2017 Lightbend Inc. <https://www.lightbend.com>
 */
package play.api.libs.ws.akkahttp

import java.net.URI

import akka.actor.ActorSystem
import akka.http.scaladsl.coding.{ Deflate, Gzip, NoCoding }
import akka.http.scaladsl.{ Http, HttpsConnectionContext }
import akka.http.scaladsl.model.HttpHeader.ParsingResult
import akka.http.scaladsl.model.Uri.Authority
import akka.http.scaladsl.model._
import akka.http.scaladsl.model.headers._
import akka.stream.Materializer
import play.api.libs.ws._

import scala.collection.immutable
import scala.concurrent.duration.Duration.Infinite
import scala.concurrent.{ Future, TimeoutException }
import scala.concurrent.duration.{ Duration, FiniteDuration }

object StandaloneAkkaHttpWSRequest {
  def apply(url: String)(implicit sys: ActorSystem, mat: Materializer, ctx: HttpsConnectionContext, config: WSClientConfig): StandaloneAkkaHttpWSRequest = {

    val requestTransformations = Seq(
      (req: HttpRequest) => config.userAgent.fold(req)(ua => req.withHeaders(`User-Agent`(ua))),
      (req: HttpRequest) => if (!config.compressionEnabled) req else req.withHeaders(`Accept-Encoding`(HttpEncodings.gzip, HttpEncodings.deflate))
    )

    new StandaloneAkkaHttpWSRequest(
      requestTransformations.foldLeft(HttpRequest().withUri(Uri.parseAbsolute(url)))((req, f) => f(req)),
      Seq.empty, Duration.Inf
    )
  }
}

final class StandaloneAkkaHttpWSRequest private (
    val request: HttpRequest,
    val filters: Seq[WSRequestFilter],
    val timeout: Duration
)(implicit
    val sys: ActorSystem,
    val mat: Materializer,
    val ctx: HttpsConnectionContext,
    val config: WSClientConfig) extends StandaloneWSRequest {

  override type Self = StandaloneWSRequest
  override type Response = StandaloneWSResponse

  // FIXME make configurable
  final val MaxRedirects = 5

  /**
   * The base URL for this request
   */
  override def url: String = request.uri.toString

  /**
   * The URI for this request
   */
  override def uri: URI =
    // TODO convert directly instead of via string
    new URI(request.uri.toString)

  /**
   * The content type for this request, if any is defined.
   */
  override def contentType: Option[String] = request.entity.contentType match {
    case ContentTypes.NoContentType => None
    case contentType => Some(contentType.toString)
  }

  /**
   * The method for this request
   */
  override def method: String = request.method.value

  /**
   * The body of this request
   */
  override def body: WSBody = request.entity match {
    case HttpEntity.Empty => EmptyBody
    case HttpEntity.Strict(_, data) => InMemoryBody(data)
    case e: HttpEntity.Chunked => SourceBody(e.dataBytes)
    case other => throw new IllegalArgumentException(s"Unknown request body type [${request.entity}]")
  }

  /**
   * The headers for this request
   */
  override def headers: Map[String, Seq[String]] =
    request.headers
      .map(h => HttpHeader.parse(h.name().toLowerCase, h.value()))
      .collect {
        case ParsingResult.Ok(header, _) => header
      }
      .groupBy(_.name)
      .mapValues(_.map(_.value()))

  /**
   * The query string for this request
   */
  override def queryString: Map[String, Seq[String]] = request.uri.query().toMultiMap

  /**
   * The cookies for this request
   */
  override def cookies: Seq[WSCookie] =
    request.cookies.map(c => DefaultWSCookie(c.name, c.value))

  /**
   * A calculator of the signature for this request
   */
  override def calc: Option[WSSignatureCalculator] =
    // FIXME https://github.com/playframework/play-ws/issues/207
    None

  /**
   * The authentication this request should use
   */
  override def auth: Option[(String, String, WSAuthScheme)] =
    request.header[Authorization].map(_.credentials).collect {
      case BasicHttpCredentials(username, password) => (username, password, WSAuthScheme.BASIC)
      case other => throw new IllegalArgumentException(s"Authorization [$other] not yet supported")
    }

  /**
   * Whether this request should follow redirects
   */
  override def followRedirects: Option[Boolean] =
    Some(config.followRedirects)

  /**
   * The timeout for the request
   */
  override def requestTimeout: Option[Int] = timeout match {
    case duration: FiniteDuration => Some(duration.toMillis.toInt)
    case duration: Infinite => None
  }

  /**
   * The virtual host this request will use
   */
  override def virtualHost: Option[String] = request.header[Host].map(h => h.host.address + ":" + h.port)

  /**
   * The proxy server this request will use
   */
  override def proxyServer: Option[WSProxyServer] =
    // FIXME https://github.com/playframework/play-ws/issues/207
    None

  /**
   * sets the signature calculator for the request
   *
   * @param calc the signature calculator
   */
  override def sign(calc: WSSignatureCalculator): Self =
    // FIXME https://github.com/playframework/play-ws/issues/207
    ???

  /**
   * sets the authentication realm
   */
  override def withAuth(username: String, password: String, scheme: WSAuthScheme): Self =
    scheme match {
      case WSAuthScheme.BASIC => copy(request = request.addHeader(Authorization(BasicHttpCredentials(username, password))))
      case authScheme => throw new IllegalArgumentException(s"Authentication scheme [$scheme] not yet supported")
    }

  /**
   * Returns this request with the given headers, discarding the existing ones.
   *
   * @param headers the headers to be used
   */
  override def withHttpHeaders(headers: (String, String)*): Self =
    copy(request = request.withHeaders(headers
      .map { case (name, value) => HttpHeader.parse(name, value) }
      .collect {
        case ParsingResult.Ok(header, _) => header
      }
      .to[immutable.Seq]
    ))

  /**
   * Returns this request with the given query string parameters, discarding the existing ones.
   *
   * @param parameters the query string parameters
   */
  override def withQueryStringParameters(parameters: (String, String)*): Self =
    copy(request = request.copy(uri = request.uri.withQuery(Uri.Query(parameters.reverse: _*))))

  /**
   * Returns this request with the given cookies, discarding the existing ones.
   *
   * @param cookies the cookies to be used
   */
  override def withCookies(cookies: WSCookie*): Self =
    copy(request = request.mapHeaders(h => h.filter {
      case c: Cookie => false
      case _ => true
    } ++ cookies.map(c => Cookie(c.name, c.value))))

  /**
   * Sets whether redirects (301, 302) should be followed automatically
   */
  override def withFollowRedirects(follow: Boolean): Self =
    copy(config = config.copy(followRedirects = follow))

  /**
   * Sets the maximum time you expect the request to take.
   * Use Duration.Inf to set an infinite request timeout.
   * Warning: a stream consumption will be interrupted when this time is reached unless Duration.Inf is set.
   */
  override def withRequestTimeout(timeout: Duration): Self =
    copy(timeout = timeout)

  /**
   * Adds a filter to the request that can transform the request for subsequent filters.
   */
  override def withRequestFilter(filter: WSRequestFilter): Self =
    copy(filters = filters :+ filter)

  /**
   * Sets the virtual host to use in this request
   */
  override def withVirtualHost(vh: String): Self =
    copy(request = request.addHeader(Host(Authority.parse(vh))))

  /**
   * Sets the proxy server to use in this request
   */
  override def withProxyServer(proxyServer: WSProxyServer): Self =
    // FIXME https://github.com/playframework/play-ws/issues/207
    ???

  /**
   * Sets the method for this request
   */
  override def withMethod(method: String): Self =
    copy(request = request.withMethod(
      HttpMethods.getForKey(method)
        .getOrElse(throw new IllegalArgumentException(s"Unknown HTTP method $method"))))

  /**
   * Sets the body for this request.
   */
  override def withBody[T: BodyWritable](body: T): Self = {
    val writable = implicitly[BodyWritable[T]]
    val contentType = ContentType.parse(writable.contentType) match {
      case Left(_) => throw new IllegalArgumentException(s"Unknown content type [${writable.contentType}]")
      case Right(ct) => ct
    }

    val requestWithEntity = request.withEntity(writable.transform(body) match {
      case InMemoryBody(bytes) => HttpEntity(contentType, bytes)
      case SourceBody(source) => HttpEntity(contentType, source)
      case EmptyBody => HttpEntity.Empty
    })

    copy(request = requestWithEntity)
  }

  /**
   * Performs a GET.
   */
  override def get(): Future[Response] =
    execute(HttpMethods.GET.value)

  /**
   * Performs a PATCH request.
   *
   * @param body the payload body submitted with this request
   * @return a future with the response for the PATCH request
   */
  override def patch[T: BodyWritable](body: T): Future[Response] =
    withBody(body).execute(HttpMethods.PATCH.value)

  /**
   * Performs a POST request.
   *
   * @param body the payload body submitted with this request
   * @return a future with the response for the POST request
   */
  override def post[T: BodyWritable](body: T): Future[Response] =
    withBody(body).execute(HttpMethods.POST.value)

  /**
   * Performs a PUT request.
   *
   * @param body the payload body submitted with this request
   * @return a future with the response for the PUT request
   */
  override def put[T: BodyWritable](body: T): Future[Response] =
    withBody(body).execute(HttpMethods.PUT.value)

  /**
   * Perform a DELETE on the request asynchronously.
   */
  override def delete(): Future[Response] =
    execute(HttpMethods.DELETE.value)

  /**
   * Perform a HEAD on the request asynchronously.
   */
  override def head(): Future[Response] =
    execute(HttpMethods.HEAD.value)

  /**
   * Perform a OPTIONS on the request asynchronously.
   */
  override def options(): Future[Response] =
    execute(HttpMethods.OPTIONS.value)

  /**
   * Executes the given HTTP method.
   *
   * @param method the HTTP method that will be executed
   * @return a future with the response for this request
   */
  override def execute(method: String): Future[Response] =
    withMethod(method).execute()

  /**
   * Execute this request
   */
  override def execute(): Future[Response] = {
    def requestToResponse(request: HttpRequest): Future[HttpResponse] =
      Http().singleRequest(request, connectionContext = ctx)

    def handleRedirects(redirectsLeft: Int, request: HttpRequest)(response: HttpResponse): Future[HttpResponse] = {
      if (redirectsLeft < 1) {
        Future.failed(new IllegalStateException("Maximum redirect reached: " + MaxRedirects))
      } else if (response.status.isRedirection() && config.followRedirects) {
        val location = response.header[Location].get.uri
        val redirectUri =
          if (location.isRelative)
            if (location.path.startsWithSlash)
              request.uri.withPath(location.path)
            else
              request.uri.withPath(request.uri.path ++ location.path)
          else location
        requestToResponse(request.withUri(redirectUri)).flatMap(handleRedirects(redirectsLeft - 1, request))(sys.dispatcher)
      } else {
        Future.successful(response)
      }
    }

    def decodeResponse(response: HttpResponse): HttpResponse = {
      val decoder = response.encoding match {
        case HttpEncodings.gzip ⇒
          Gzip
        case HttpEncodings.deflate ⇒
          Deflate
        case _ ⇒
          NoCoding
      }
      decoder.decodeMessage(response)
    }

    val akkaExecutor = WSRequestExecutor { request =>
      import sys.dispatcher
      val akkaRequest = request.asInstanceOf[StandaloneAkkaHttpWSRequest].request
      val timeoutFuture = timeout match {
        case duration: Infinite => Seq.empty[Future[StandaloneAkkaHttpWSResponse]]
        case duration: FiniteDuration =>
          Seq(akka.pattern.after(duration, sys.scheduler)(Future.failed(new TimeoutException(s"Request timeout after $duration"))))
      }
      Future.firstCompletedOf(
        timeoutFuture :+
          requestToResponse(akkaRequest)
          .flatMap(handleRedirects(MaxRedirects, akkaRequest))
          .map(decodeResponse)
          .map(StandaloneAkkaHttpWSResponse.apply)(sys.dispatcher)
      )
    }

    val execution = filters.foldRight(akkaExecutor)((filter, executor) => filter.apply(executor))

    execution(this)
  }

  /**
   * Execute this request and stream the response body.
   */
  override def stream(): Future[Response] = execute()

  private def copy(
    request: HttpRequest = request,
    filters: Seq[WSRequestFilter] = filters,
    timeout: Duration = timeout,
    config: WSClientConfig = config
  ) = new StandaloneAkkaHttpWSRequest(request, filters, timeout)(sys, mat, ctx, config)
}

File: play-akka-http-ws-standalone/src/main/java/play/libs/ws/akkahttp/StandaloneAkkaHttpWSRequest.java
/*
 * Copyright (C) 2009-2017 Lightbend Inc. <https://www.lightbend.com>
 */
package play.libs.ws.akkahttp;

import akka.actor.ActorSystem;
import akka.http.impl.model.parser.HeaderParser$;
import akka.http.javadsl.Http;
import akka.http.javadsl.HttpsConnectionContext;
import akka.http.javadsl.model.*;
import akka.http.javadsl.model.headers.Authorization;
import akka.http.javadsl.model.headers.BasicHttpCredentials;
import akka.http.javadsl.model.headers.Cookie;
import akka.http.javadsl.model.headers.UserAgent;
import akka.http.scaladsl.model.headers.ProductVersion;
import akka.japi.Pair;
import akka.parboiled2.ParserInput$;
import akka.pattern.PatternsCS;
import akka.stream.Materializer;
import play.api.libs.ws.WSClientConfig;
import play.libs.ws.*;
import scala.concurrent.duration.FiniteDuration;
import scala.util.Either;

import java.nio.charset.Charset;
import java.time.Duration;
import java.util.*;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.stream.Collectors;

public final class StandaloneAkkaHttpWSRequest implements StandaloneWSRequest {

  private final HttpRequest request;
  private final List<WSRequestFilter> filters;
  private final Duration timeout;

  private final ActorSystem sys;
  private final Materializer mat;
  private final HttpsConnectionContext ctx;
  private final WSClientConfig config;

  StandaloneAkkaHttpWSRequest(String url, ActorSystem sys, Materializer mat, HttpsConnectionContext ctx, WSClientConfig config) {
    this(HttpRequest.create(url), new ArrayList<>(), Duration.ZERO, sys, mat, ctx, config);
  }

  private StandaloneAkkaHttpWSRequest(HttpRequest request, List<WSRequestFilter> filters, Duration timeout, ActorSystem sys, Materializer mat, HttpsConnectionContext ctx, WSClientConfig config) {
    this.request = config.userAgent().fold(
      () -> request,
      (ua) -> request.addHeader(UserAgent.create(
        // FIXME JAVA API expose ProductVersion.parseMultiple in Java API
        scala.collection.JavaConverters.seqAsJavaList(ProductVersion.parseMultiple(ua)).toArray(new ProductVersion[0])
      ))
    );
    this.filters = filters;
    this.timeout = timeout;
    this.sys = sys;
    this.mat = mat;
    this.ctx = ctx;
    this.config = config;
  }

  /**
   * Perform a GET on the request asynchronously.
   *
   * @return a promise to the response
   */
  @Override
  public CompletionStage<? extends StandaloneWSResponse> get() {
    return execute(HttpMethods.GET.value());
  }

  /**
   * Perform a PATCH on the request asynchronously.
   *
   * @param body the BodyWritable
   * @return a promise to the response
   */
  @Override
  public CompletionStage<? extends StandaloneWSResponse> patch(BodyWritable body) {
    return setBody(body).execute(HttpMethods.PATCH.value());
  }

  /**
   * Perform a POST on the request asynchronously.
   *
   * @param body the BodyWritable
   * @return a promise to the response
   */
  @Override
  public CompletionStage<? extends StandaloneWSResponse> post(BodyWritable body) {
    return setBody(body).execute(HttpMethods.POST.value());
  }

  /**
   * Perform a PUT on the request asynchronously.
   *
   * @param body the BodyWritable
   * @return a promise to the response
   */
  @Override
  public CompletionStage<? extends StandaloneWSResponse> put(BodyWritable body) {
    return setBody(body).execute(HttpMethods.PUT.value());
  }

  /**
   * Perform a DELETE on the request asynchronously.
   *
   * @return a promise to the response
   */
  @Override
  public CompletionStage<? extends StandaloneWSResponse> delete() {
    return execute(HttpMethods.DELETE.value());
  }

  /**
   * Perform a HEAD on the request asynchronously.
   *
   * @return a promise to the response
   */
  @Override
  public CompletionStage<? extends StandaloneWSResponse> head() {
    return execute(HttpMethods.HEAD.value());
  }

  /**
   * Perform an OPTIONS on the request asynchronously.
   *
   * @return a promise to the response
   */
  @Override
  public CompletionStage<? extends StandaloneWSResponse> options() {
    return execute(HttpMethods.OPTIONS.value());
  }

  /**
   * Executes an arbitrary method on the request asynchronously.
   *
   * @param method The method to execute
   * @return a promise to the response
   */
  @Override
  public CompletionStage<? extends StandaloneWSResponse> execute(String method) {
    return setMethod(method).execute();
  }

  /**
   * Executes an arbitrary method on the request asynchronously.  Should be used with setMethod().
   *
   * @return a promise to the response
   */
  @Override
  public CompletionStage<? extends StandaloneWSResponse> execute() {
    final WSRequestExecutor akkaExecutor = (request) -> {
      final CompletableFuture<StandaloneAkkaHttpWSResponse> resultFuture =
        Http.get(sys)
          .singleRequest(((StandaloneAkkaHttpWSRequest)request).request, ctx)
          .thenApply((r) -> new StandaloneAkkaHttpWSResponse(r, mat))
          .toCompletableFuture();

      if (timeout.equals(Duration.ZERO)) {
        return resultFuture.thenApply((response) -> (StandaloneWSResponse)response);
      }
      else {
        final CompletableFuture<StandaloneAkkaHttpWSResponse> timeoutException = new CompletableFuture<>();
        timeoutException.completeExceptionally(new TimeoutException("Request timeout after " + timeout));

        final CompletableFuture<StandaloneAkkaHttpWSResponse> timeoutFuture =
          PatternsCS.after(
            FiniteDuration.apply(timeout.toNanos(), TimeUnit.NANOSECONDS),
            sys.scheduler(), sys.dispatcher(),
            timeoutException
          ).toCompletableFuture();

        return CompletableFuture.anyOf(resultFuture, timeoutFuture).thenApply((response) -> (StandaloneWSResponse)response);
      }
    };

    return filters.stream()
      .reduce(akkaExecutor, (executor, filter) -> filter.apply(executor), (exec1, exec2) -> exec2)
      .apply(this);
  }

  /**
   * Executes this request and streams the response body.
   * <p>
   * Use {@code response.bodyAsSource()} with this method.
   *
   * @return a promise to the response
   */
  @Override
  public CompletionStage<? extends StandaloneWSResponse> stream() {
    return execute();
  }

  /**
   * Sets the HTTP method this request should use, where the no args execute() method is invoked.
   *
   * @param method the HTTP method.
   * @return the modified WSRequest.
   */
  @Override
  public StandaloneWSRequest setMethod(String method) {
    return copy(request.withMethod(
      HttpMethods.lookup(method)
        .orElseThrow(() -> new IllegalArgumentException("Unknown HTTP method " + method))));
  }

  /**
   * Set the body this request should use.
   *
   * @param body the body of the request.
   * @return the modified WSRequest.
   */
  @Override
  public StandaloneWSRequest setBody(BodyWritable body) {
    if (body instanceof InMemoryBodyWritable) {
      final InMemoryBodyWritable writable = (InMemoryBodyWritable)body;
      return copy(request.withEntity(
        HttpEntities.create(parseContentType(body.contentType()), writable.body().get())));
    }
    else if (body instanceof SourceBodyWritable) {
      final SourceBodyWritable writable = (SourceBodyWritable)body;
      return copy(request.withEntity(
        HttpEntities.create(parseContentType(body.contentType()), writable.body().get())));
    }
    else {
      throw new IllegalArgumentException("Unsupported BodyWritable: " + body);
    }
  }

  /**
   * Set headers to the request.  Note that duplicate headers are allowed
   * by the HTTP specification, and removing a header is not available
   * through this API. Any existing header will be discarded here.
   *
   * @param headers the headers
   * @return the modified WSRequest.
   */
  @Override
  public StandaloneWSRequest setHeaders(Map<String, List<String>> headers) {
    // FIXME JAVA API no Java Api to replace headers on a request
    HttpRequest requestNoHeaders = request;
    for (HttpHeader h: request.getHeaders()) {
      requestNoHeaders = requestNoHeaders.removeHeader(h.name());
    }

    return copy(
      requestNoHeaders.addHeaders(
        headers.entrySet().stream().flatMap(
          (h) -> h.getValue().stream().map((v) -> parseHeader(h.getKey(), v))
        ).collect(Collectors.toList()))
    );
  }

  /**
   * Adds a header to the request.  Note that duplicate headers are allowed
   * by the HTTP specification, and removing a header is not available
   * through this API. Existent headers will be preserved.
   *
   * @param name  the header name
   * @param value the header value
   * @return the modified WSRequest.
   */
  @Override
  public StandaloneWSRequest addHeader(String name, String value) {
    return copy(request.addHeader(parseHeader(name, value)));
  }

  /**
   * Sets the query string to query.
   *
   * @param query the fully formed query string
   * @return the modified WSRequest.
   */
  @Override
  public StandaloneWSRequest setQueryString(String query) {
    return copy(
      request.withUri(
        request.getUri().query(
          Query.create(query))));
  }

  /**
   * Adds a query parameter with the given name, this can be called repeatedly and will preserve existing values.
   * Duplicate query parameters are allowed.
   *
   * @param name  the query parameter name
   * @param value the query parameter value
   * @return the modified WSRequest.
   */
  @Override
  public StandaloneWSRequest addQueryParameter(String name, String value) {
    return copy(
      request.withUri(
        request.getUri().query(
          request.getUri().query().withParam(name, value))));
  }

  /**
   * Sets the query string parameters. This will discard existing values.
   *
   * @param params the query string parameters
   * @return the modified WSRequest.
   */
  @Override
  public StandaloneWSRequest setQueryString(Map<String, List<String>> params) {
    return copy(
      request.withUri(
        request.getUri().query(
          Query.create(params.entrySet().stream().flatMap(
            (entry) -> entry.getValue().stream().map((value) -> Pair.create(entry.getKey(), value))
          ).collect(Collectors.toList())))));
  }

  /**
   * Add a new cookie. This can be called repeatedly and will preserve existing cookies.
   *
   * @param cookie the cookie to be added
   * @return the modified WSRequest.
   * @see #addCookies(WSCookie...)
   * @see #setCookies(List)
   */
  @Override
  public StandaloneWSRequest addCookie(WSCookie cookie) {
    return addCookies(cookie);
  }

  /**
   * Add new cookies. This can be called repeatedly and will preserve existing cookies.
   *
   * @param cookies the list of cookies to be added
   * @return the modified WSRequest.
   * @see #addCookie(WSCookie)
   * @see #setCookies(List)
   */
  @Override
  public StandaloneWSRequest addCookies(WSCookie... cookies) {
    return copy(request.addHeaders(
      Arrays.stream(cookies)
        .map((c) -> Cookie.create(c.getName(), c.getValue()))
        .collect(Collectors.toList())));
  }

  /**
   * Set the request cookies. This discard the existing cookies.
   *
   * @param cookies the cookies to be used.
   * @return the modified WSRequest.
   */
  @Override
  public StandaloneWSRequest setCookies(List<WSCookie> cookies) {
    return copy(request.removeHeader("Cookie")).addCookies(cookies.toArray(new WSCookie[cookies.size()]));
  }

  /**
   * Sets the authentication header for the current request using BASIC authentication.
   *
   * @param userInfo a string formed as "username:password".
   * @return the modified WSRequest.
   */
  @Override
  public StandaloneWSRequest setAuth(String userInfo) {
    final String[] credentials = userInfo.split(":");
    return setAuth(credentials[0], credentials[1]);
  }

  /**
   * Sets the authentication header for the current request using BASIC authentication.
   *
   * @param username the basic auth username
   * @param password the basic auth password
   * @return the modified WSRequest.
   */
  @Override
  public StandaloneWSRequest setAuth(String username, String password) {
    return setAuth(username, password, WSAuthScheme.BASIC);
  }

  /**
   * Sets the authentication header for the current request.
   *
   * @param username the username
   * @param password the password
   * @param scheme   authentication scheme
   * @return the modified WSRequest.
   */
  @Override
  public StandaloneWSRequest setAuth(String username, String password, WSAuthScheme scheme) {
    if (scheme.equals(WSAuthScheme.BASIC)) {
      return copy(request.addCredentials(BasicHttpCredentials.createBasicHttpCredentials(username, password)));
    }
    else {
      throw new IllegalArgumentException("Authentication scheme [" + scheme + "] not yet supported");
    }
  }

  /**
   * Sets an (OAuth) signature calculator.
   *
   * @param calculator the signature calculator
   * @return the modified WSRequest
   */
  @Override
  public StandaloneWSRequest sign(WSSignatureCalculator calculator) {
    // FIXME https://github.com/playframework/play-ws/issues/207
    throw new UnsupportedOperationException("Implementation is missing");
  }

  /**
   * Sets whether redirects (301, 302) should be followed automatically.
   *
   * @param followRedirects true if the request should follow redirects
   * @return the modified WSRequest
   */
  @Override
  public StandaloneWSRequest setFollowRedirects(boolean followRedirects) {
    // FIXME https://github.com/playframework/play-ws/issues/207
    throw new UnsupportedOperationException("Implementation is missing");
  }

  /**
   * Sets the virtual host as a "hostname:port" string.
   *
   * @param virtualHost the virtual host
   * @return the modified WSRequest
   */
  @Override
  public StandaloneWSRequest setVirtualHost(String virtualHost) {
    // FIXME JAVA API missing Host.create(Authority) Java Api in Akka Http
    return copy(request.addHeader(akka.http.scaladsl.model.headers.Host.apply(
      akka.http.scaladsl.model.Uri.Authority$.MODULE$.parse(
        ParserInput$.MODULE$.apply(virtualHost),
        Charset.forName("UTF8"),
        akka.http.scaladsl.model.Uri$ParsingMode$Relaxed$.MODULE$))));
  }

  /**
   * Sets the request timeout duration. Java {@link Duration} class does not have a specific instance
   * to represent an infinite timeout, but according to the docs, in practice, you can somehow emulate
   * it:
   * <p>
   * <blockquote>
   * A physical duration could be of infinite length. For practicality, the duration is stored
   * with constraints similar to Instant. The duration uses nanosecond resolution with a maximum
   * value of the seconds that can be held in a long. This is greater than the current estimated
   * age of the universe.
   * </blockquote>
   * <p>
   * Play WS uses the convention of setting a duration with negative value to have an infinite timeout.
   * So you will have:
   * <p>
   * <pre>java.time.Duration timeout = Duration.ofSeconds(-1);</pre>.
   *
   * In practice, you can also have an extreme long duration, like:
   *
   * <pre>java.time.Duration timeout = Duration.ofMillis(Long.MAX_VALUE);</pre>
   *
   * And, as the {@link Duration} docs states, this will be good enough since this duration is greater than
   * the current estimate age of the universe.
   *
   * @param timeout the request timeout in milliseconds. A duration of -1 indicates an infinite request timeout.
   * @return the modified WSRequest.
   */
  @Override
  public StandaloneWSRequest setRequestTimeout(Duration timeout) {
    return copy(timeout);
  }

  /**
   * Adds a request filter.
   *
   * @param filter a transforming filter.
   * @return the modified request.
   */
  @Override
  public StandaloneWSRequest setRequestFilter(WSRequestFilter filter) {
    final List<WSRequestFilter> newFilters = new ArrayList<>(filters);
    newFilters.add(filter);
    return copy(newFilters);
  }

  /**
   * Set the content type.  If the request body is a String, and no charset parameter is included, then it will
   * default to UTF-8.
   *
   * @param contentType The content type
   * @return the modified WSRequest
   */
  @Override
  public StandaloneWSRequest setContentType(String contentType) {
    // FIXME JAVA API entity.withContentType is missing in Java Api of Akka Http
    return copy(request.withEntity(
      ((RequestEntity)((akka.http.scaladsl.model.HttpEntity)request.entity()).withContentType((akka.http.scaladsl.model.ContentType)parseContentType(contentType)))
    ));
  }

  /**
   * @return the URL of the request.  This has not passed through an internal request builder and so will not be signed.
   */
  @Override
  public String getUrl() {
    return request.getUri().query(Query.EMPTY).toString();
  }

  /**
   * @return the headers (a copy to prevent side-effects). This has not passed through an internal request builder and so will not be signed.
   */
  @Override
  public Map<String, List<String>> getHeaders() {
    final Map<String, List<String>> headers = new HashMap<>();
    for (final HttpHeader header: request.getHeaders()) {
      if (headers.containsKey(header.name())) {
        headers.get(header.name()).add(header.value());
      }
      else {
        headers.put(header.name(), new ArrayList<>(Collections.singletonList(header.value())));
      }
    }
    return headers;
  }

  /**
   * Get all the values of header with the specified name. If there are no values for
   * the header with the specified name, than an empty List is returned.
   *
   * @param name the header name.
   * @return all the values for this header name.
   */
  @Override
  public List<String> getHeaderValues(String name) {
    return getHeaders().getOrDefault(name, new ArrayList<>());
  }

  /**
   * Get the value of the header with the specified name. If there are more than one values
   * for this header, the first value is returned. If there are no values, than an empty
   * Optional is returned.
   *
   * @param name the header name
   * @return the header value
   */
  @Override
  public Optional<String> getHeader(String name) {
    return getHeaderValues(name).stream().findFirst();
  }

  /**
   * @return the query parameters (a copy to prevent side-effects). This has not passed through an internal request builder and so will not be signed.
   */
  @Override
  public Map<String, List<String>> getQueryParameters() {
    return request.getUri().query().toMultiMap();
  }

  /**
   * @return the auth username, null if not an authenticated request.
   */
  @Override
  public String getUsername() {
    return request.getHeader(Authorization.class)
      .filter((a) -> a.credentials() instanceof BasicHttpCredentials)
      .map((a) -> (BasicHttpCredentials)a.credentials())
      .map(BasicHttpCredentials::username)
      .orElseGet(() -> null);
  }

  /**
   * @return the auth password, null if not an authenticated request
   */
  @Override
  public String getPassword() {
    return request.getHeader(Authorization.class)
      .filter((a) -> a.credentials() instanceof BasicHttpCredentials)
      .map((a) -> (BasicHttpCredentials)a.credentials())
      .map(BasicHttpCredentials::password)
      .orElseGet(() -> null);
  }

  /**
   * @return the auth scheme, null if not an authenticated request.
   */
  @Override
  public WSAuthScheme getScheme() {
    return request.getHeader(Authorization.class)
      .filter((a) -> a.credentials() instanceof BasicHttpCredentials)
      .map((a) -> WSAuthScheme.BASIC)
      .orElseGet(() -> null);
  }

  /**
   * @return the signature calculator (example: OAuth), null if none is set.
   */
  @Override
  public WSSignatureCalculator getCalculator() {
    // FIXME https://github.com/playframework/play-ws/issues/207
    throw new UnsupportedOperationException("Implementation is missing");
  }

  /**
   * Gets the original request timeout duration, passed into the request as input.
   *
   * @return the timeout duration.
   */
  @Override
  public Duration getRequestTimeoutDuration() {
    return timeout;
  }

  /**
   * @return true if the request is configure to follow redirect, false if it is configure not to, null if nothing is configured and the global client preference should be used instead.
   */
  @Override
  public boolean getFollowRedirects() {
    // FIXME https://github.com/playframework/play-ws/issues/207
    return false;
  }

  /**
   * @return the content type, if any, or null.
   */
  @Override
  public String getContentType() {
    // FIXME JAVA API no CotentTypes.NoContentType Java Api in Akka Http
    return request.entity().getContentType()
      .equals(akka.http.scaladsl.model.ContentTypes.NoContentType()) ? null : request.entity().getContentType().toString();
  }

  private HttpHeader parseHeader(String name, String value) {
    // FIXME JAVA API missing HttpHeader.parse Java API in Akka Http
    final akka.http.scaladsl.model.HttpHeader.ParsingResult result =
      akka.http.scaladsl.model.HttpHeader$.MODULE$.parse(name, value, HeaderParser$.MODULE$.DefaultSettings());

    if (result instanceof akka.http.scaladsl.model.HttpHeader$ParsingResult$Ok) {
      return ((akka.http.scaladsl.model.HttpHeader$ParsingResult$Ok)result).header();
    }
    else {
      throw new IllegalArgumentException("Unable to parse header [" + name + "] with value [" + value + "]");
    }
  }

  private ContentType parseContentType(String contentType) {
    // FIXME JAVA API missing ContentType.parse Java API in Akka Http
    final Either<scala.collection.immutable.List<akka.http.scaladsl.model.ErrorInfo>, akka.http.scaladsl.model.ContentType> contentTypeEither = akka.http.scaladsl.model.ContentType$.MODULE$.parse(contentType);
    if (contentTypeEither.isRight()) {
      return contentTypeEither.right().get();
    }
    else {
      throw new IllegalArgumentException("Unable to parse content type: " + contentType);
    }
  }

  private StandaloneWSRequest copy(HttpRequest request) {
    return new StandaloneAkkaHttpWSRequest(request, this.filters, this.timeout, this.sys, this.mat, this.ctx, this.config);
  }

  private StandaloneWSRequest copy(List<WSRequestFilter> filters) {
    return new StandaloneAkkaHttpWSRequest(this.request, filters, this.timeout, this.sys, this.mat, this.ctx, this.config);
  }

  private StandaloneWSRequest copy(Duration timeout) {
    return new StandaloneAkkaHttpWSRequest(this.request, this.filters, timeout, this.sys, this.mat, this.ctx, this.config);
  }

}
